### DISCUSSION TITLE
How to deal with "refusing to merge unrelated histories" error

### DISCUSSION THREAD

**Discussion by ewan-delanoy** [community/community#22075] [New to GitHub]
<p>Short version of my question : For years, I have been using a simple, single one-branch, one-contributor public online Github repo. A few days ago my computer died suddenly and I bought a new one. Now Github refuses to connect the local repo<br>
from my new computer to the online repo saying “refusing to merge unrelated histories”. What do I do ?</p>
<p>Long version of my question : here is the exact sequence of commands that I did in my new computer :</p>
<p>Step 1 : Download git, use <code>git config</code> to tell git about my username &amp; email. Worked fine.</p>
<p>All the following command were executed in my new local repo’s main directory</p>
<p>Step 2 : <code>git init</code>. Worked fine.</p>
<p>Step 3 : Do a <code>git add</code> on all my files. Worked fine.</p>
<p>Step 4 : Do the first commit : <code>git commit -m "First commit from new computer"</code>. Worked fine.</p>
<p>Step 5 : Do <code>git remote add origin [https://github.com/roparzhhemon/myremoterepo.git](https://github.com/roparzhhemon/myremoterepo.git)</code>. Worked fine, according to<br>
<code>git remote -v</code>.</p>
<p>Step 6 : <code>git push</code>. Got the following error message :</p>
<pre><code>fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use
git push --set-upstream origin master
</code></pre>
<p>Step 7 : Do as I’m told, and type : <code>git push --set-upstream origin master</code>. Got the following error message :</p>
<pre><code>error: failed to push some refs to [remote repo]
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
</code></pre>
<p>Step 8 : Do as I’m told, and type : <code>git pull</code>. Got the following message :</p>
<pre><code>warning: no common commits
remote: Counting objects: 11450, done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 11450 (delta 13), reused 17 (delta 8), pack-reused 11425
Receiving objects: 100% (11450/11450), 1.96 MiB | 2.65 MiB/s, done.
Resolving deltas: 100% (7710/7710), done.
From [remote repo]
* [new branch] master -&gt; origin/master
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.
git pull &lt;remote&gt; &lt;branch&gt;
If you wish to set tracking information for this branch you can do so with:
git branch --set-upstream-to=origin/&lt;branch&gt; master
</code></pre>
<p>Step 9 : Do as I’m told, and type : <code>git branch --set-upstream-to=origin/master master</code>. Seemed to work, output the following : </p>
<pre><code>Branch 'master' set up to track remote branch 'master' from 'origin'.
</code></pre>
<p>Step 10 : Try <code>git pull</code> again. Got the error message :</p>
<pre><code>fatal: refusing to merge unrelated histories
</code></pre>
<p>Step 11 : Try <code>git push</code> again. Got the error message :</p>
<pre><code>To [remote repo]
! [rejected] master -&gt; master (non-fast-forward)
error: failed to push some refs to [remote repo]
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
</code></pre>

**Comment by mpboom** [comment]
<p>If you just start a new repository, the history will be completely different than GitHub knows it.</p>
<p>Use</p>
<pre><code>git clone &lt;repo&gt;
</code></pre>
<p>and just make commits on top of that. Then push.</p>

**Comment by xvrtzn** [comment, marked as answer]
<p>I ended in the same situation after a series of mishapps. Have a look at <a href="https://stackoverflow.com/questions/37937984/git-refusing-to-merge-unrelated-histories-on-rebase" rel="nofollow noopener">this thread</a> : I just did a </p>
<pre><code>git pull --allow-unrelated-histories
</code></pre>
<p>and it was fixed!</p>

**Reply by edu18RR** [reply]
Gracias por la ayuda, me sirvio de mucho.

**Comment by geetika96** [comment]
<p> I was also facing the same problem. But I realised I was trying to push my one git project material on other remote git project. Solution:</p>
<p>Clone the fresh project again and this time make sure cloning link is starting from ssh if you have added the SSH key and https:// if you have added the puttyKey.</p>

**Comment by surendra-ids** [comment]
<p>Thanks, worked fine for me.</p>

**Comment by SimpleSamples** [comment]
<p>I am in a similar situation where I am getting the same messages. Probably I just need to do the</p>
<pre><code>--allow-unrelated-histories
</code></pre>
<p>But I wish I knew what situations where I should not. Obviously we get the message because there might sometimes be a good reason to <strong>not</strong> do the allow. I got into my situation because I did not know what I was doing and did something I should not have.</p>

**Comment by anu-una** [comment]
<p>I am receiving the message saying “there is no tracking information for the current branch”</p>

**Comment by iofirag** [comment]
<p>Help for me. thanks!</p>

**Comment by haxianhe** [comment]
<p>Thank, you for you help.</p>

**Comment by imisamarti** [comment]
<pre><code>git pull --allow-unrelated-histories
</code></pre>
<p>worked</p>

**Reply by BaoZhuhan** [reply]
thank you.

**Comment by ClauGarfiasMx** [comment]
<p>This is great, thank you very much!</p>
<p><a href="http://gph.is/1U7OwpV" rel="nofollow noopener">http://gph.is/1U7OwpV</a></p>

**Comment by sawepeter** [comment]
<p>This command is a dangerous one. “git pull --allow-unrelated-histories” it has messed up my whole codebase i dont know what happened and am not in a position to recover from it</p>

**Comment by farney9** [comment]
<p>It’s worked for me.<br>
Very thanks! i have solved my problem</p>

**Comment by Aisuko** [comment]
<p>Thanks, It works for me.</p>

**Comment by shubhangi-tandon** [comment]
<p>Works for me too ! thank you :slight_smile:</p>

**Comment by sudora73** [comment]
<p>Thanks, Woked superbly for me as well.<br>
but little more specific:</p>
<p>git pull &lt;remote&gt; &lt;branch&gt; --allow-unrelated-histories</p>

**Comment by BonnieSawe** [comment]
<p>Worked like a charm</p>

**Comment by jpinkham1** [comment]
<p>I got this error, but found another cause:</p>
<p>somehow, the ssh url of another project got into this one’s .git/config file.</p>
<p>So before running this to accept ‘unrelated histories’, you may want to confirm you have the right repo’s url.</p>
<p>I fixed mine and it started working again.  :slight_smile:</p>

**Comment by auspicious99** [comment]
<p>Probably because you did a <code>git init</code> on something like “the same” set of files locally, then tried to merge with the files on the repository. The better way is to clone from the repository, then work from that.</p>

**Comment by ghost** [comment]
<p>I tried the --allow-unrelated-histories and got this:</p>
<p>warning: Cannot merge binary files: biostats.final.3.2.pdf (HEAD vs. 93379526ed908abd4f58b2010eb055aace5df522)</p>
<p>warning: Cannot merge binary files: .DS_Store (HEAD vs. 93379526ed908abd4f58b2010eb055aace5df522)</p>
<p>Auto-merging &lt;file name here&gt;</p>
<p>CONFLICT (add/add): Merge conflict in &lt;file name here&gt;</p>
<p>Auto-merging &lt;file name here&gt;</p>
<p>CONFLICT (add/add): Merge conflict in  &lt;file name here&gt;</p>
<p>Auto-merging  &lt;file name here&gt;</p>
<p>CONFLICT (add/add): Merge conflict in  &lt;file name here&gt;</p>
<p>Auto-merging  &lt;file name here&gt;</p>
<p>CONFLICT (add/add): Merge conflict in  &lt;file name here&gt;</p>
<p>Auto-merging  &lt;file name here&gt;</p>
<p>Automatic merge failed; fix conflicts and then commit the result.</p>
